### ISSUE TITLE
support reproducible Python builds

### ISSUE THREAD

**Issue by 05ed1d68-9f72-4552-af38-41e9d578a41d** [python/cpython#73894]
BPO | [29708](https://bugs.python.org/issue29708)
--- | :---
Nosy | @warsaw, @vstinner, @ericvsmith, @benjaminp, @mcepl, @merwok, @methane, @zooba, @dstufft, @bmwiedemann, @FRidh, @commodo, @mingwandroid, @eli-schwartz, @miss-islington, @jefferyto, @obfusk
PRs | <li>python/cpython#296</li><li>python/cpython#5200</li><li>python/cpython#5306</li><li>python/cpython#5313</li><li>python/cpython#8057</li><li>python/cpython#9607</li><li>python/cpython#10775</li>
Files | <li>[python39_2.html](https://bugs.python.org/file49708/python39_2.html "Uploaded as text/plain at 2020-12-30.13:08:45 by @FRidh"): Python 3.9.1 diffoscope report</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-03-03.11:36:06.114>
labels = ['build', '3.10']
title = 'support reproducible Python builds'
updated_at = <Date 2021-04-22.17:01:17.438>
user = 'https://github.com/bmwiedemann'
```

bugs.python.org fields:
```python
activity = <Date 2021-04-22.17:01:17.438>
actor = 'obfusk'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Build']
creation = <Date 2017-03-03.11:36:06.114>
creator = 'bmwiedemann'
dependencies = []
files = ['49708']
hgrepos = []
issue_num = 29708
keywords = ['patch']
message_count = 40.0
messages = ['288880', '288883', '288889', '288948', '301354', '309394', '309395', '309401', '309870', '309905', '309931', '309972', '310010', '310012', '310292', '310636', '310637', '310652', '310661', '311317', '313312', '313313', '313383', '313384', '313391', '320942', '320989', '321002', '327480', '330623', '347971', '384065', '384066', '384099', '384100', '384104', '384110', '384123', '386272', '391616']
nosy_count = 18.0
nosy_names = ['barry', 'vstinner', 'eric.smith', 'benjamin.peterson', 'mcepl', 'eric.araujo', 'sascha_silbe', 'methane', 'steve.dower', 'dstufft', 'bmwiedemann', 'Frederik Rietdijk', 'Alexandru Ardelean', 'Ray Donnelly', 'eschwartz', 'miss-islington', 'jefferyto', 'obfusk']
pr_nums = ['296', '5200', '5306', '5313', '8057', '9607', '10775']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue29708'
versions = ['Python 3.10']
```

</p></details>

**Comment by 05ed1d68-9f72-4552-af38-41e9d578a41d** [comment]
See https://reproducible-builds.org/ and https://reproducible-builds.org/docs/buy-in/ for why this is a good thing to have in general.

Fedora, openSUSE and possibly other Linux distributions package .pyc files as part of their binary rpm packages and they are not trivial to drop [1].

A .pyc header includes the timestamp of the source .py file
which creates non-reproducible builds when the .py file is touched during build time (e.g. for a version.py).
As of 2017-02-10 in openSUSE Factory this affected 476 packages (such as python-amqp and python3-Twisted).


[1] http://lists.opensuse.org/opensuse-packaging/2017-02/msg00086.html

**Comment by ericvsmith** [comment]
--
Eric.

**Comment by warsaw** [comment]
Shouldn't this at least also cover Python 3.7?  And should it be officially backported?  I would think that if https://github.com/python/cpython/pull/296 gets accepted for 3.7, then distros that care can cherry pick it back into whatever versions they still support.  It probably needn't be officially cherry picked upstream.

(FWIW, this doesn't affect the Debian ecosystem since we don't ship pycs in debs.)

**Comment by 05ed1d68-9f72-4552-af38-41e9d578a41d** [comment]
backports are optional.
It can help reduce duplicated work for the various distributions.
Currently, I think master and 2.7 are the most relevant targets.

**Comment by benjaminp** [comment]
I have proposed PEP-552 to address this issue.

**Comment by AraHaan** [comment]
For me, I find that manually rewriting my py files in C using the C python api made my builds reproducible (I think) as well as faster.

**Comment by alittlesir** [comment]
https://github.com/python/cpython/issues/93317 The bug I encountered was not caused by the timestamp, I have tools that control timestamps to keep timestamps consistent.

if i close to  compiling of multithreading, it don't appear again. I think it's caused by these characters, these strings are ”{“ ”{{“ ”}“ ”}“

![image](https://user-images.githubusercontent.com/67537842/171371684-b4c4f648-3c5a-4395-a9db-6ec862993fe5.png)

![image](https://user-images.githubusercontent.com/67537842/171371717-1dd2882c-9f0a-421b-bee8-39f0f2c9e6d0.png)

There may be DA/FA differences in some pyc files in the python standard library. If two pythons with differences in the standard library pyc are used to compile the same py file containing the "{" string, the pyc generated will also have DA/FA differences. This is because some pyc files in the standard library will be automatically loaded when the python -m compileall xxx command is executed. When the pyc file is loaded, it recognizes that TYPE_SHORT_ASCII_INTERNED will intern the corresponding string, and marshal.c can see this operation , so when new pyc is produced, "{" will also be written as TYPE_SHORT_ASCII_INTERNED type.

**Comment by alittlesir** [comment]
> Thank you for reporting. It is other source of undeterministic pyc as you said.
> 
> Single codepoint strings (and empty string) are shared among interpreter. "{" is not name, so compile() doesn't intern it. But since it is singleton, same instance is shared from many places.
> 
> Most simple solution is interning all singletons in `_PyRuntimeState_Init`.

Do you have any plans to fix this bug in the near future?

**Comment by methane** [comment]
> Do you have any plans to fix this bug in the near future?

I will fix it before 3.12 become beta if no one fix it until then.

**Comment by AraHaan** [comment]
I wish 3.12 was 4.0 so that way the 2to3 would get nuked.

**Comment by vstinner** [comment]
> I wish 3.12 was 4.0 so that way the 2to3 would get nuked. But most likely 4.0 will land in 2050.

2to3 is deprecated since Python 3.11 and so can be removed in Python 3.13 (or later): https://github.com/python/cpython/issues/84540 It's a problem with the parser implementation in lib2to3 which doesn't support Python 3.10 grammar.

**Comment by AraHaan** [comment]
What if the entire python standard libraries was to be rewritten in C? I just thought of this and I think it would help eliminate this issue as well as make the interpreter faster because then it can be further optimized.

**Comment by ericvsmith** [comment]
> What if the entire python standard libraries was to be rewritten in C?

This is such an enormous task that it will never happen. Plus it would make CPython harder to maintain, and make it harder to share anything in the stdlib with other implementations.

**Comment by arhadthedev** [comment]
> What if the entire python standard libraries was to be rewritten in C

Lots of pointer management boilerplate, plus Python operators will inflate into lenghty function calls.

**Comment by arhadthedev** [comment]
> Alternatively python could build in cython into it’s compile() function under python to compile scripts into pyd’s. However the generated c code would have to be deleted to not waste a ton of disk space.

Using Cython as an optional feature enabled in `configure` and `PCBuild` looks promising.

However, we need a person who is ready to add the lines into `configure.ac`, `Makefile.pre.in`, `PCBuild/get_externals.bat`, relevant Visual Studio projects, and possibly submit necessary patches to Cython to support such a use case.

**Comment by alittlesir** [comment]
If compiled with PGO twice, it will also produce inconsistencies, and there are many different places.

What's the reason for this

**Comment by methane** [comment]
> If compiled with PGO twice, it will also produce inconsistencies, and there are many different places.

I don't understand what you are saying.
Please provide concrete and complete step to reproduce.

**Comment by mcepl** [comment]
> > If compiled with PGO twice, it will also produce inconsistencies, and there are many different places.
> 
> I don't understand what you are saying. Please provide concrete and complete step to reproduce.

I don’t understand why https://github.com/python/cpython/issues/93317 didn’t get linked here.

**Comment by merwok** [comment]
It was, but the UI is not very obvious: https://github.com/python/cpython/issues/73894#ref-issue-1251584367

**Comment by compete2cooperate** [comment]
I am a novice on python compilation and trying determininstic pyc. As simplest use case, I cloned git repo with a sample .py file in two places (src1 and src2) inside the same build environment (docker container) and compiled both of them using invalidation_mode=CHECKED_HASH and SOURCE_DATE_EPOCH set. But I can see that hash of corresponding .pyc files are different. (source file inside both src1 and src2 have same hash but different file attributes notably mtime)

How can I compile source file in src1 and src2 so that correponding pyc have same hash? I am using python version 3.10.12. My apologies if this is not the right place to ask.

**Comment by manueljacob** [comment]
It seems like the main issue has been solved. Would it make sense to open separate issues for the remaining specific problems and close this one?

**Comment by picnixz** [comment]
Yes, let's do that instead of having one big issue that is quite old.
